### PR TITLE
test(auth): added first tests for authentication

### LINKED ISSUE #281: Add missing e2e and unit tests for `auth` command
Prerequisites
- [x] #270 

Goal: working on core e2e workflows + unit tests

### PR DESCRIPTION
**Pull request by ctrlc03**
Implemented a number of test cases for authentication using GitHub OAuth2 for prod testing and username and password for the Firebase emulator. 
Update: due to some errors with GitHub and Puppeteer, prod testing include more reliable tests with username and password. This is documented on #288 and #282.

fix #281

### CODE REVIEW

File: packages/actions/test/e2e/00-authentication.test.ts [hunk 1]
```diff
@@ -56,24 +62,47 @@ describe("Authentication", () => {
...
+    it("should not be possible to authenticate with an incorrect email", async () => {
+        await expect(signInWithEmailAndPassword(userAuth, "wrongEmail", userPassword)).to.be.rejected
+    })
 
-    it("should not be possible to authenticate if the user has been disabled from the Authentication service by coordinator", async () => {})
+    it("should not be possible to authenticate if Firebase is unreachable", async () => {
+        // @todo mock unreachable firebase.
+    })
+    it("should not be possible to authenticate twice", async () => {
+        // @todo Implement checks to prevent double authentication.
```
Comment (0xjei) [inline packages/actions/test/e2e/00-authentication.test.ts:95]: I think double authentication is currently avoided in the `auth` command (by checking local access token). Hmm, idk if there's a better way using Firebase methods (e.g., you got recognize if already auth and try to do it twice).

File: packages/actions/test/utils/authentication.ts [hunk 2]
```diff
@@ -195,6 +195,279 @@ export const simulateOnVerification = async (verification: Verification): Promis
     ])
 }
 
+/**
+ * Simulate callback to cancel authorization for Github OAuth2.0 device flow.
+ * @param verification <Verification> - the data from Github OAuth2.0 device flow.
+ */
```
Comment (0xjei) [inline packages/actions/test/utils/authentication.ts:201]: Add @todo to be completed after solving issue #289 et al

File: packages/actions/test/utils/authentication.ts [hunk 3]
```diff
@@ -195,6 +195,279 @@ export const simulateOnVerification = async (verification: Verification): Promis
...
+    await Promise.all([await continueButton?.click(), ghPage.waitForNavigation()])
+
+    await browser.close()
+}
+
+/**
+ * Simulate callback for unreachable GitHub website for Github OAuth2.0 device flow.
+ * @param verification <Verification> - the data from Github OAuth2.0 device flow.
+ */
+/* eslint-disable @typescript-eslint/no-unused-vars */
```
Comment (0xjei) [inline packages/actions/test/utils/authentication.ts:422]: Same for others as above

File: packages/actions/test/utils/index.ts [hunk 4]
```diff
@@ -13,5 +13,8 @@ export {
     createNewFirebaseUserWithEmailAndPw,
     getLastGithubVerificationCode,
     simulateOnVerification,
-    authenticateUserWithGithub
+    authenticateUserWithGithub,
```
Comment (0xjei) [inline packages/actions/test/utils/index.ts:16]: Avoid exporting since we are not going to use them rn

**Review by 0xjei — CHANGES_REQUESTED** [review]
Looks got, few nits and considerations.

File: packages/actions/test/e2e/00-authentication.test.ts (hunk 1 above)
Comment (ctrlc03) [reply packages/actions/test/e2e/00-authentication.test.ts:95]:
```js
await signInWithEmailAndPassword(userAuth, user.data.email, userPassword)
await signInWithEmailAndPassword(userAuth, user.data.email, userPassword)
```

Authenticating twice with email and password does not give any error. This test might have to be removed from the actions package and be part of possible tests for the cli itself.

Comment (0xjei) [reply packages/actions/test/e2e/00-authentication.test.ts:98]: Cool, thanks for finding this out. So, if aren't any errors from Firebase methods, we could assume to move to the CLI (where is already covered).

**Review by 0xjei — APPROVED** [review]
:cool:

**Pull request merged**